### PR TITLE
Module 3

### PR DESCRIPTION
**Pull request by jianingxu1**
The objective was to experiment with and implement a linear model capable of predicting product outcomes in a production environment.

I conducted experiments with various linear models, adjusting hyperparameters such as penalty and regularization, as well as exploring different sets of features for training. The best-performing model identified in the notebook was then implemented into production within the .py files.

Due to time constraints, I did not delve into categorical values. However, I provided a brief explanation of my approach to addressing them. Additionally, I conducted only a single test due to time limitations but would like to know which functions would be important to test in this project.

Finally, while the overall structure of the solution draws inspiration from class, I tried to adapt it and independently implement each step to gain a thorough understanding of the process. For the next modules, I will try to PR a solution before class.

### CODE REVIEW

File: src/module_3/module_3_model.md
Comment (sergiorozada12) [inline src/module_3/module_3_model.md:1]: I leave you here some comments about the NB:
* Good summary of the insights
* Good check the assert to prevent from data leakage
* Consider add ```;``` to ```lr_lasso.fit(X_train[train_cols], y_train)``` to avoid printing garbage
* Consider normalizing the coefficients to compare them 
* As a follow up: what happens if you retrain the model after prunning the irrelevant features? Can you get rid of more features?
* Good proposed follow up with the categorical encoding

File: src/module_3/data_processing.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+def filter_orders_with_minimum_size(
+    df: pd.DataFrame, min_size: int = 5
+) -> pd.DataFrame:
+    """
+    Filter orders that have at least 5 products
+    """
+    filt = df.query("outcome == 1.0").groupby("order_id").size() >= min_size
+    order_ids = set(filt[filt].index)
+    filtered_df = df[df["order_id"].isin(order_ids)]
```
Comment (sergiorozada12) [inline src/module_3/data_processing.py:28]: Have you considered doing a join instead isin? Are they the same from a computational perspective? If you want to dig more into this, consider the same case with a big datased in pyspark, and try to read about it.

File: src/module_3/data_processing.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    return filtered_df
+
+
+def build_feature_frame(data: pd.DataFrame) -> pd.DataFrame:
+    logging.info("Building feature frame")
+    preprocessed_data = (
+        data.pipe(filter_orders_with_minimum_size, min_size=5)
```
Comment (sergiorozada12) [inline src/module_3/data_processing.py:35]: Another folow up, how pipe compares to filter in this case, or assign to apply? Nits for this use case, interesting if you have this kind of details in your head.

File: src/module_3/inference.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
+import logging
```
Comment (sergiorozada12) [inline src/module_3/inference.py:1]: Good use of logging

File: src/module_3/train.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    test_proportion: float,
+):
+    """
+    Split train and test by date as dataset contains timeseries
+    """
+    train_proportion = 1 - test_proportion
+
+    df_sorted = df.sort_values(by="order_date")
+
+    # Consider only orders and not pairs of order, item
```
Comment (sergiorozada12) [inline src/module_3/train.py:31]: Is this comment needed?

File: src/module_3/train.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
+import os
+import logging
+import pandas as pd
+import joblib
+from datetime import datetime
+
+# from collections import defaultdict
```
Comment (sergiorozada12) [inline src/module_3/train.py:7]: If you don't use them, remove them

File: src/module_3/train.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+    # Train the model with entire dataset
+    model = PushModel(
+        penalty_param=PENALTY_PARAM,
+        c_param=C_PARAM,
+        prediction_threshold=PREDICTION_THRESHOLD,
+    )
+    model.fit(data)
+
+    # # predictions = infer("hi", "hi").to_dict()
```
Comment (sergiorozada12) [inline src/module_3/train.py:69]: Same here, the PR cannot have non-useful code

File: tests/module_3/test_data_processing.py [hunk 7]
```diff
@@ -0,0 +1,17 @@
+import pandas as pd
+
+from src.module_3.data_processing import filter_orders_with_minimum_size
+
+
+def test_orders_with_minimum_size() -> None:
```
Comment (sergiorozada12) [inline tests/module_3/test_data_processing.py:6]: Can you think of other tests to add here?

**Review by sergiorozada12 — APPROVED** [review]
Good PR overall, you can merge it. I know that time is a hard constrain, but tests are critical. Between adding a couple of follow-ups (pruning features to retrain) or making the tests, go for the tests.
